src/MediaWiki/Search/Search.php

```diff
@@ -163,6 +104,52 @@ private function searchFallbackSearchEngine( $term, $fulltext ) {
 	}
 
 	/**
+	 * @return SearchEngine
+	 */
+	public function getFallbackSearchEngine() {
+		if ( $this->fallbackSearch === null ) {
+			$type = ApplicationFactory::getInstance()->getSettings()->get( 'smwgFallbackSearchType' );
+
+			$dbr = $this->getDB();
+
+			if ( $type === null ) {
+				$type = $this->getDefaultSearchEngineTypeForDB( $dbr );
+			}
+
+			$this->assertIsSearchEngineClassName( $type );
+
+			$this->fallbackSearch = new $type( $dbr );
+		}
+
+		return $this->fallbackSearch;
+	}
+
+	/**
+	 * @param DatabaseBase $db
+	 * @return string
+	 */
+	private function getDefaultSearchEngineTypeForDB( DatabaseBase $db ) {
+		if ( method_exists( 'SearchEngineFactory', 'getSearchEngineClass' ) ) { // MW > 1.27
+			return \SearchEngineFactory::getSearchEngineClass( $db );
+		}
+
+		return $db->getSearchEngine();
+	}
+
+	/**
+	 * @param string $className
+	 */
+	private function assertIsSearchEngineClassName( $className ) {
+		if ( !class_exists( $className ) ) {
+			throw new RuntimeException( "$className does not exist." );
+		}
+
+		if ( !is_subclass_of( $className, 'SearchEngine' ) || $className === 'SMWSearch' ) {
```

Comment:
It is intentional. If the fallback search is again an SWMSearch we run into a recursion.